presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.zorder;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class ZOrderIntegers
+{
+    private ZOrderIntegers() {}
+
+    // TODO: make int and long z-address version
+    public static byte[] encode(List<Integer> input, List<Integer> bitRanges)
+            throws Exception
+    {
+        if (input.size() > 10) {
+            throw new Exception("Z-ordering does not support more than 10 input numbers");
+        }
+
+        // Transfer into array for faster access and to change values
+        int[] inputList = new int[input.size()];
+        for (int index = 0; index < input.size(); index++) {
+            inputList[index] = input.get(index);
+            if (inputList[index] < 0) {
+                throw new Exception("Negative numbers are not supported for z-ordering");
+            }
+        }
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : bitRanges) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int inputBitPos = 0; inputBitPos < maxBitLength; inputBitPos++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (inputBitPos >= bitRanges.get(index)) {
+                    continue;
+                }
+                int maskedBit = (inputList[index] >> (bitRanges.get(index) - inputBitPos - 1)) & 1;
+                address[bitIndex >> 3] |= maskedBit << bitOffset;
+                bitIndex--;
+                bitOffset = bitIndex & 7;
+            }
+        }
+        return address;
+    }
+
+    public static List<Integer> decode(List<Integer> bitRanges, byte[] address)
```

Comment:
Same as above. bitRanges should be calculated by the library; the signature will be simple: ... decode(byte[] address).

This could introduce the state for the object: bitRanges will be a state of the object. When a user create a ZOrderIntegers to encode, they have to use the same object to decode. Therefore, your two functions can not be static anymore.